sql/byte_buffer.go

```diff
@@ -0,0 +1,72 @@
+package sql
+
+import (
+	"sync"
+)
+
+var SingletonBuf = NewByteBuffer(16000)
+
+var defaultByteBuffCap = 1000
+
+var ByteBufPool = sync.Pool{
+	New: func() any {
+		// The Pool's New function should generally only return pointer
+		// types, since a pointer can be put into the return interface
+		// value without an allocation:
+		return NewByteBuffer(defaultByteBuffCap)
+	},
+}
+
+type ByteBuffer struct {
```

Comment:
I agree with all of this, but there are two caveats that limit our ability to let the runtime handle this for us. (1) The runtime chooses doubling based on the cap of the slice, not the full backing array. So for the current setup, the doubled array is usually actually smaller than the original backing array. (2) Doubled arrays are not reference swapped, we need a handle to the new buffer to know when to grow.

I'm not aware of how to override the default runtime growth behavior to ignore the slice cap and instead double based on the backing array cap. So`BytesBuffer` still does the doubling, and a `Grow(n int)` interface to track when this should happen. We pay for 2 mallocs on doubling, because the first one is never big enough. Not calling `Grow` after allocing, or growing by too small of length compared to the allocations used will stomp previously written memory.